### ISSUE TITLE
Warn about expensive queries in GithubAuthV4Service

### ISSUE THREAD

**Issue by chris48s** [badges/shields#3810]
:clipboard: **Description**

We should use this method
https://developer.github.com/v4/guides/resource-limitations/#calculating-nodes-in-a-call
to work out the query cost before we make the call and error if it is going to be over some threshold instead of issueing the query. Because we're using `gql` template literals, we've got the parsed query available but there's some work to accurately implement the calculation.

Just thinking one step ahead, if there isn't already a third party package we can lean on here, this might be a good project to develop as a package. As GH V4 API usage increases, I can see this being a common requirement for other applications using the API.

See notes: https://github.com/badges/shields/pull/3763#discussion_r306513358

**Comment by paulmelnikow** [comment]
Ah, interesting, I didn't realize we could compute the cost before making the call. Neat!

**Comment by calebcartwright** [comment]
I've not had a chance to review the new BaseGraphqlService implementation so apologies if this question is off base, but would the query (and the associated cost) be static with each service classe based on BaseGraphql, or will the cost vary based on the user-provided parameters? 

Asked differently, would the cost calculation be something that we'd need to do dynamically at runtime or more statically against each service class?

**Comment by chris48s** [comment]
Yeah its not completely trivial. Having thought about it more, I think it has to be a runtime check. A graphql query is just a string, so we could do

```js

var number_of_repos = 3;
var query = `query {
  viewer {
    name
     repositories(last: ${number_of_repos}) {
       nodes {
         name
       }
     }
   }
}`
query = gql(query)
```


but also, the number of nodes can depend on the gql variables, so

```gql
query($number_of_repos:Int!) {
  viewer {
    name
     repositories(last: $number_of_repos) {
       nodes {
         name
       }
     }
   }
}
variables {
   "number_of_repos": 3
}
```

is another way to do that too. My hope is that mostly we're not going to be doing a lot of that, but a general purpose implementation would need to account for these things.
